src/main/java/ai/promoted/delivery/client/DeliveryRequestState.java

```diff
@@ -0,0 +1,69 @@
+package ai.promoted.delivery.client;
+
+import java.util.List;
+import ai.promoted.delivery.model.Insertion;
+import ai.promoted.delivery.model.Request;
+import ai.promoted.delivery.model.Response;
+
+/**
+ * Holds state of a request during processing.
+ */
+class DeliveryRequestState {
+
+  /** The delivery request we're processing */
+  private final DeliveryRequest deliveryRequest;
+
+  /** Any insertions on the request that were not passed to Delivery API */
+  private List<Insertion> nonApiInsertions;
+  
+  /**
+   * Creates a state object for the given request.
+   * 
+   * @param deliveryRequest the delivery request we are processing.
+   */
+  public DeliveryRequestState(DeliveryRequest deliveryRequest) {
+    this.deliveryRequest = deliveryRequest;
+  }
+  
+  /**
+   * Gets the response to return to the client with any necessary post-processing.
+   * @param resp the response from Delivery API.
+   * @return the post-processed response
+   */
+  public Response getResponseToReturn(Response resp) {
+    
+    // Add any insertions back past the maximum, and assign them an insertion id.
+    if (nonApiInsertions != null) {
+      int nextIndex = 0;
+      if (resp.getInsertion().size() > 0) {
```

Comment:
I don't think this works with paging.

Reply:
I think it does (I unit-tested with a simulated non-zero offset) because it starts numbering positions on the added-back insertions starting at +1 the position of the last returned insertion (which is the largest since they always come back sorted).

Reply:
Now I see what you're saying.